src/test/groovy/com/datadog/ddwaf/WafHandleTest.groovy

```
@@ -3,228 +3,180 @@
     }
 
     @Test
-    void 'throw an exception when both persistent and ephemeral are null in wafContext'() {
-        ctx = new WafHandle('test', null, ARACHNI_ATOM_V2_1)
-        wafContext = ctx.openContext()
-        metrics = ctx.createMetrics()
-        shouldFail(AbstractWafException) {
-            wafContext.run(null, null, limits, metrics)
-        }
+    void 'waf handle has known actions'() {
+        wafDiagnostics = builder.addOrUpdateConfig('test', ARACHNI_ATOM_BLOCK)
+        handle = builder.buildWafHandleInstance()
+
+        String[] actions = handle.knownActions
+        assert actions != null
+        assert actions.length > 0
     }
 
     @Test
-    void 'throw an exception when both persistent and ephemeral are null'() {
-        ctx = new WafHandle('test', null, ARACHNI_ATOM_V2_1)
-        metrics = ctx.createMetrics()
-        shouldFail(AbstractWafException) {
-            ctx.runRules(null, limits, metrics)
+    void 'knownAddresses throws exception after handle is closed'() {
+        wafDiagnostics = builder.addOrUpdateConfig('test', ARACHNI_ATOM_V1_0)
+        handle = builder.buildWafHandleInstance()
+        handle.close()
+
+        def exception = shouldFail(IllegalStateException) {
+            handle.knownAddresses
         }
+        assert exception.message == 'This WafHandle is no longer online'
     }
 
     @Test
-    void 'constructor throws if given a null context'() {
-        shouldFail(NullPointerException) {
-            new WafContext(null)
+    void 'getKnownActions throws exception after handle is closed'() {
+        wafDiagnostics = builder.addOrUpdateConfig('test', ARACHNI_ATOM_BLOCK)
+        handle = builder.buildWafHandleInstance()
+        handle.close()
+
+        def exception = shouldFail(IllegalStateException) {
+            handle.knownActions
         }
+        assert exception.message == 'This WafHandle is no longer online'
     }
 
     @Test
-    void 'should throw if double free'() {
```

Should be restored.

---

waf handle can be closed multiple times safely is the new one - sending a destroy to ddwaf does not cause an issue, i guess i should throw an exception java side? confirm?

---

It might be a good idea to not throw on close, but to avoid so many moving pieces both here and at dd-trace-java, we might want to keep the old behavior (throw on close if it was already closed).